mlflow/R/mlflow/R/tracking-runs.R

```diff
@@ -116,48 +116,40 @@ mlflow_get_run <- function(run_id = NULL, client = NULL) {
 #'   data may be written.
 #' @template roxlate-client
 #' @template roxlate-run-id
-#' @param metrics A named list of metrics to log.
-#' @param params A named list of params to log.
-#' @param tags A named list of tags to log.
-#' @param timestamps (Optional) A list of timestamps of the same length as `metrics`.
+#' @param metrics A dataframe of metrics to log, containing the following columns: "key", "value",
+#'  "step", "timestamp"
+#' @param params A dataframe of params to log, containing the following columns: "key", "value"
+#' @param tags A dataframe of tags to log, containing the following columns: "key", "value"
 #' @export
-mlflow_log_batch <- function(metrics = NULL, params = NULL, tags = NULL, timestamps = NULL,
-                             run_id = NULL, client = NULL) {
-  c(client, run_id) %<-% resolve_client_and_run_id(client, run_id)
+mlflow_log_batch <- function(metrics = NULL, params = NULL, tags = NULL, run_id = NULL,
+                             client = NULL) {
+  validate_batch_input_column_names("metrics", metrics,
+                                    c("key", "value", "step", "timestamp"))
+  validate_batch_input_column_names("params", params, c("key", "value"))
+  validate_batch_input_column_names("tags", tags, c("key", "value"))
 
-  metrics <- construct_batch_list(metrics)
-  params <- construct_batch_list(params)
-  tags <- construct_batch_list(tags)
-
-  if (!is.null(metrics)) {
-    metrics <- if (is.null(timestamps)) {
-      purrr::map(metrics, ~ c(.x, timestamp = current_time()))
-    } else {
-      if (length(metrics) != length(timestamps))
-        stop("`metrics` and `timestamps` must be of the same length.", call. = FALSE)
-      timestamps <- purrr::map(timestamps, ~ list(timestamp = .x))
-      purrr::map2(metrics, timestamps, c)
-    }
-  }
+  c(client, run_id) %<-% resolve_client_and_run_id(client, run_id)
 
   mlflow_rest("runs", "log-batch", client = client, verb = "POST", data = list(
     run_id = run_id,
     metrics = metrics,
     params = params,
     tags = tags
   ))
-
   invisible(NULL)
 }
 
-construct_batch_list <- function(l) {
-  if (is.null(l)) {
-    l
-  } else {
-    l %>%
-      purrr::imap(~ list(key = .y, value = .x)) %>%
-      unname()
+validate_batch_input_column_names <- function(input_type, input_dataframe, expected_column_names) {
```

Comment:
I think we should consider how we handle missing values for each of the column types. E.g. if step is missing the step recorded will be 0 which I think is what we want. But what if timestamp is missing? Should we fill in the current time in that case?

And should we fail if key or value columns have missing values?

Reply:
Per offline discussion, I've updated input validation logic to throw an exception if any missing (`NA`) entry is detected.